nightly.swift

```diff
@@ -2,13 +2,166 @@
+
+enum SourceHost: String {
+    case GitHub = "github.com"
+}
+
+// MARK: - Error
+
+enum ValidatorError: Error {
+    case invalidURL(String)
+    case timedOut
+    case noData
+    case networkingError(Error)
+    case decoderError(Error)
+    case unknownGitHost(String?)
+    case fileSystemError(Error)
+    case badPackageDump(String?)
+    case missingProducts
+    case rateLimitExceeded(Int)
+    case packageDoesNotExist(String)
+    case dumpTimedOut
+    case unknownError(Error)
+    case repoIsFork
+    case outdatedToolsVersion
+    
+    var localizedDescription: String {
+        switch self {
+        case .invalidURL(let url):
+            return "Invalid URL: \(url)"
+        case .timedOut:
+            return "Request Timed Out"
+        case .noData:
+            return "No Data Received"
+        case .dumpTimedOut:
+            return "Dump Timed Out"
+        case .networkingError(let error), .decoderError(let error), .fileSystemError(let error), .unknownError(let error):
+            return error.localizedDescription
+        case .unknownGitHost(let host):
+            return "Unknown URL host: \(host ?? "nil")"
+        case .badPackageDump(let output):
+            return "Bad Package Dump -- \(output ?? "No Output")"
+        case .missingProducts:
+            return "Missing Products"
+        case .rateLimitExceeded(let limit):
+            return "Rate Limit of \(limit) Exceeded"
+        case .packageDoesNotExist(let url):
+            return "Package Does Not Exist: \(url)"
+        case .repoIsFork:
+            // A decision has been made that the index as a whole should support forks, but not as part of dependency analysis.
+            //
+            // This is because there's an unhealthy amount of forks with a single patch to simply make the dependency work with their library,
+            // thse are often unmaintained and don't delivery huge amounts of value.
+            return "Forks are not added as part of dependency analysis"
+        case .outdatedToolsVersion:
+            return "Repo is using an outdated package format"
+        }
+    }
+}
+
+// MARK: - Networking
+
+func downloadSync(url: String, timeout: Int = 10) -> Result<Data, ValidatorError> {
+    let semaphore = DispatchSemaphore(value: 0)
+    
+    guard let apiURL = URL(string: url) else {
+        return .failure(.invalidURL(url))
+    }
+    
+    var payload: Data?
+    var taskError: ValidatorError?
+    
+    var request = URLRequest(url: apiURL)
+    
+    if let token = patToken, apiURL.host?.contains(SourceHost.GitHub.rawValue) == true {
+        request.addValue("Basic \(token)", forHTTPHeaderField: "Authorization")
+    }
+    
+    let task = session.dataTask(with: request) { (data, response, error) in
+        
+        let httpResponse = response as? HTTPURLResponse
+        
+        if let limit = httpResponse?.value(forHTTPHeaderField: "X-RateLimit-Limit").flatMap(Int.init),
```

Comment:
This isn't working 100% of the time, we should do a contains on the `message` on "API rate limit exceeded". This'll reduce more of the noise.

> {"message":"API rate limit exceeded for 199.7.166.17. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)","documentation_url":"https://developer.github.com/v3/#rate-limiting"}